assembly/std/arguments.ts

```diff
@@ -0,0 +1,165 @@
+  }
+
+  /**
+   *
+   * @return {string} the string
+   */
+  nextString(): string {
+    let offset: i32 = this.offset as i32;
+    const length = u8(this.argsString.codePointAt(offset));
+    const result = this.argsString.slice(offset + 1, offset + length + 1);
+    this.offset = offset + length + 1;
+    return result;
+  }
+
+  /**
+   * @return {u64}
+   */
+  nextU64(): u64 {
+    const byteArray = this.fromByteString(this.argsString);
+    const amount = this.toU64(byteArray, this.offset as u8);
+    this.offset += 8;
+    return amount;
+  }
+
+  /**
+   * @return {i64}
+   */
+  nextI64(): i64 {
+    const byteArray = this.fromByteString(this.argsString);
+    const amount = this.toI64(byteArray, this.offset as u8);
+    this.offset += 8;
+    return amount;
+  }
+
+  // Setter
+
+  /**
+   *
+   * @param {T} arg the argument to add
+   *
+   * @return {Args} the modified Arg instance
+   */
+  add<T>(arg: T): Args {
+    if (arg instanceof Address) {
+      this.argsString = this.argsString.concat(arg.toStringSegment());
+    } else if (arg instanceof String || typeof arg == 'string') {
+      const str: string = arg.toString();
+      this.argsString = this.argsString.concat(
+        String.fromCharCode(u8(str.length)).concat(str as string),
+      );
+    } else if (arg instanceof u64) {
+      this.argsString = this.argsString.concat(
+        ByteArray.fromU64(arg as u64).toByteString(),
+      );
+    } else if (arg instanceof i64 || typeof arg == 'number') {
+      this.argsString = this.argsString.concat(
+        ByteArray.fromI64(arg as i64).toByteString(),
+      );
+    } else {
+      throw new Error('unsupported type: ' + typeof arg);
+    }
+
+    return this;
+  }
+
+  // Utils
+
+  /**
+   *
+   * @param {string} byteString
+   * @return {Uint8Array}
+   */
+  private fromByteString(byteString: string): Uint8Array {
+    const byteArray = new Uint8Array(byteString.length);
+    for (let i = 0; i < byteArray.length; i++) {
+      byteArray[i] = u8(byteString.charCodeAt(i));
+    }
+    return byteArray;
+  }
+
+  /**
+   *
+   * @param {Uint8Array} byteArray
+   * @param {u8} offset
+   * @return {u64}
+   */
+  private toU64(byteArray: Uint8Array, offset: u8 = 0): u64 {
+    if (byteArray.length - offset < 8) {
+      return <u64>NaN;
+    }
+
+    let x: u64 = 0;
+    x = (x | byteArray[offset + 7]) << 8;
+    x = (x | byteArray[offset + 6]) << 8;
+    x = (x | byteArray[offset + 5]) << 8;
+    x = (x | byteArray[offset + 4]) << 8;
+    x = (x | byteArray[offset + 3]) << 8;
+    x = (x | byteArray[offset + 2]) << 8;
+    x = (x | byteArray[offset + 1]) << 8;
+    x = x | byteArray[offset + 0];
+    return x;
+  }
+
+  /**
+   * @param {Uint8Array} byteArray
+   * @param {u8} offset
+   * @return {i64}
+   */
+  private toI64(byteArray: Uint8Array, offset: u8 = 0): i64 {
+    if (byteArray.length - offset < 8) {
+      return <i64>NaN;
+    }
+
+    let x: i64 = 0;
+    x = (x | byteArray[offset + 7]) << 8;
```

Comment:
So the toU64() does the same thing as toI64() ?
I think you should add Unit tests to test these functions.
and if its really the same processing inside, factorize the code.

Personnally i would to it like this (not tested)
x= x | byteArray[offset] as u64
x= x | ( byteArray[offset + 1] as u64 >> 1)
x= x | ( byteArray[offset + 2) as u64 >> 2)

Reply:
I understood the logic 🙌
and it seems this code is already tested in some way
Code factorization will be nice but not mandatory.
So i will cancel my change request !